File: app/src/main/java/ru/magflayer/spectrum/presentation/pages/main/MainPresenter.java
package ru.magflayer.spectrum.presentation.pages.main;

import com.arellomobile.mvp.InjectViewState;
import com.squareup.otto.Subscribe;

import javax.inject.Inject;

import ru.magflayer.spectrum.domain.entity.event.FabClickEvent;
import ru.magflayer.spectrum.domain.injection.InjectorManager;
import ru.magflayer.spectrum.domain.manager.CameraManager;
import ru.magflayer.spectrum.presentation.common.android.navigation.router.MainRouter;
import ru.magflayer.spectrum.presentation.common.model.PageAppearance;
import ru.magflayer.spectrum.presentation.common.mvp.BasePresenter;

@InjectViewState
public class MainPresenter extends BasePresenter<MainView> {

    @Inject
    CameraManager cameraManager;
    @Inject
    MainRouter mainRouter;

    @Override
    protected void inject() {
        InjectorManager.getAppComponent().inject(this);
    }

    @Override
    public void attachView(final MainView view) {
        super.attachView(view);
        try {
            cameraManager.open();
        } catch (RuntimeException e) {
            logger.error("Camera not available: ", e);
        }
    }

    @Override
    public void detachView(final MainView view) {
        try {
            cameraManager.close();
        } catch (RuntimeException e) {
            logger.error("Cannot close stop camera");
        }
        super.detachView(view);
    }

    void openMainScreen() {
        mainRouter.openCameraScreen();
    }

    void handleFabClick() {
        bus.post(new FabClickEvent());
    }

    @Subscribe
    public void onPageAppearanceChanged(final PageAppearance pageAppearance) {
        if (pageAppearance.getShowFloatingButton() != null) {
            getViewState().showFloatingButton(pageAppearance.getShowFloatingButton());
        }
    }
}

File: app/src/main/java/ru/magflayer/spectrum/presentation/pages/main/MainPresenter.kt
package ru.magflayer.spectrum.presentation.pages.main

import moxy.InjectViewState
import ru.magflayer.spectrum.domain.interactor.PageAppearanceInteractor
import ru.magflayer.spectrum.presentation.common.model.PageAppearance
import ru.magflayer.spectrum.presentation.common.mvp.BasePresenter
import javax.inject.Inject

@InjectViewState
class MainPresenter @Inject constructor(
    private val pageAppearanceInteractor: PageAppearanceInteractor
) : BasePresenter<MainView>() {

    override fun onFirstViewAttach() {
        execute(
            pageAppearanceInteractor.observePageAppearance(),
            onSuccess = { pageAppearance -> handlePageAppearanceChanged(pageAppearance) }
        )
    }

    fun handleFabClick() {
        pageAppearanceInteractor.publishFabEvent()
    }

    private fun handlePageAppearanceChanged(pageAppearance: PageAppearance) {
        when (pageAppearance.floatingButtonState) {
            PageAppearance.FloatingButtonState.VISIBLE -> {
                viewState.showFloatingButton(true)
            }
            PageAppearance.FloatingButtonState.INVISIBLE -> {
                viewState.showFloatingButton(false)
            }
            else -> {
                /* do nothing */
            }
        }
    }
}
